### PR TITLE
Added forgot email functionality and a discussion forum for the trip

### PR DESCRIPTION
**Pull request by runarf**
Fixed the translations and added elements id for testing. Did a rebase for the first time (Y). There are some really old commits included but they don't do any damage, and I don't know how to remove them... Seems like the travis is OK with the build.

### CODE REVIEW

File: app/controllers/comments_controller.rb [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  end
+
+  def create
+    authorize @trip, :show?
+    @comment = @discussion.comments.build(comment_params)
+    @comment.user = current_user
+    respond_to do |format|
+      if @comment.save
+        format.html {
+          flash[:notice] = "Comment has been created."
```
Comment (ritualz) [inline app/controllers/comments_controller.rb]: Translation

File: app/controllers/comments_controller.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+          redirect_to [@trip, @discussion] }
+        format.js {
+          # look at view/comment/create.js.erb
+        }
+      else
+        format.html {
+          flash[:alert] = "Comment has not been created"
```
Comment (ritualz) [inline app/controllers/comments_controller.rb]: Translation

File: app/controllers/comments_controller.rb [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  end
+
+  def edit
+    authorize @trip, :update?
+  end
+
+  def update
+    authorize @trip, :update?
+    if @comment.update(comment_params)
+      flash[:notice] = "Comment has been updated."
```
Comment (ritualz) [inline app/controllers/comments_controller.rb]: Translation

File: app/controllers/comments_controller.rb [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      redirect_to [@trip, @discussion]
+    else
+      flash[:alert] = "Comment has not been updated."
```
Comment (ritualz) [inline app/controllers/comments_controller.rb]: Translation

File: app/controllers/comments_controller.rb [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      render :edit
+    end
+  end
+
+  def destroy
+    authorize @trip, :update?
+    @comment.destroy
+    flash[:notice] = "Comment has been deleted."
```
Comment (ritualz) [inline app/controllers/comments_controller.rb]: Translation

File: app/views/comments/_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+<%= form_for [@trip, @discussion, @comment] do |f| %>
+    <% if @comment.errors.any? %>
+        <div id="error_explanation">
+          <h2><%= pluralize(@comment.errors.count, "error") %> prohibited this comment from being saved:</h2>
```
Comment (ritualz) [inline app/views/comments/_form.html.erb]: Translation

File: app/views/comments/edit.html.erb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<% content_for :page_title do %>
+    "Edit comment"
```
Comment (ritualz) [inline app/views/comments/edit.html.erb]: Translation

File: app/views/comments/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+<% content_for :page_title do %>
+    "New comment"
```
Comment (ritualz) [inline app/views/comments/new.html.erb]: Translation

File: app/views/devise/passwords/edit.html.erb [hunk 9]
```diff
@@ -1,22 +1,40 @@
-<h2>Change your password</h2>
-
-<%= form_for(resource, as: resource_name, url: password_path(resource_name), html: { method: :put }) do |f| %>
-  <%= devise_error_messages! %>
-  <%= f.hidden_field :reset_password_token %>
+<% content_for :page_title do %>
+    <i class="fa fa-history"></i> <strong>Reset password</strong>
```
Comment (ritualz) [inline app/views/devise/passwords/edit.html.erb]: Translation

File: app/views/devise/passwords/edit.html.erb [hunk 10]
```diff
@@ -1,22 +1,40 @@
...
+<% end %>
 
-  <div class="field">
-    <%= f.label :password, "New password" %><br />
-    <%= f.password_field :password, autofocus: true, autocomplete: "off" %>
+<div class="block animation-fadeInQuickInv">
+  <div class="block-title">
+    <div class="block-options pull-right">
+      <%= link_to new_user_session_path, title: "Back to login", class: "btn btn-effect-ripple btn-primary", data: {toggle: "tooltip", placement: "left"} do %>
```
Comment (ritualz) [inline app/views/devise/passwords/edit.html.erb]: Translation

File: app/views/devise/passwords/edit.html.erb [hunk 11]
```diff
@@ -1,22 +1,40 @@
...
+          <i class="fa fa-user"></i>
+      <% end %>
+    </div>
+    <h2>Reset</h2>
```
Comment (ritualz) [inline app/views/devise/passwords/edit.html.erb]: Translation

File: app/views/devise/passwords/edit.html.erb [hunk 12]
```diff
@@ -1,22 +1,40 @@
...
+      <% end %>
+      <%= f.hidden_field :reset_password_token %>
 
-  <div class="actions">
-    <%= f.submit "Change my password" %>
-  </div>
-<% end %>
+      <div class="form-group">
+        <div class="col-xs-12">
+          <%= f.password_field :password, placeholder: "New password", class: "form-control", autofocus: true, autocomplete: "off" %>
```
Comment (ritualz) [inline app/views/devise/passwords/edit.html.erb]: Translation

File: app/views/devise/passwords/edit.html.erb [hunk 13]
```diff
@@ -1,22 +1,40 @@
...
+        </div>
+      </div>
+
+      <div class="form-group">
+        <div class="col-xs-12">
+          <%= f.password_field :password_confirmation, placeholder: "Confirm new password", class: "form-control", autocomplete: "off" %>
```
Comment (ritualz) [inline app/views/devise/passwords/edit.html.erb]: Translation

File: app/views/devise/passwords/edit.html.erb [hunk 14]
```diff
@@ -1,22 +1,40 @@
...
+        </div>
+      </div>
 
-<%= render "devise/shared/links" %>
+      <div class="form-group form-actions">
+        <div class="col-xs-12 text-right">
+          <%= f.submit "Change my password", class: "btn btn-effect-ripple btn-sm btn-primary" %>
```
Comment (ritualz) [inline app/views/devise/passwords/edit.html.erb]: Translation

File: app/views/discussions/edit.html.erb [hunk 15]
```diff
@@ -0,0 +1,5 @@
+<% content_for :page_title do %>
+    "Edit discussion"
```
Comment (ritualz) [inline app/views/discussions/edit.html.erb]: Translation

File: app/views/discussions/index.html.erb [hunk 16]
```diff
@@ -0,0 +1,29 @@
+<% content_for :page_title do %>
+    "Discussions"
```
Comment (ritualz) [inline app/views/discussions/index.html.erb]: Translation

File: app/views/discussions/new.html.erb [hunk 17]
```diff
@@ -0,0 +1,5 @@
+<% content_for :page_title do %>
+    "New discussion"
```
Comment (ritualz) [inline app/views/discussions/new.html.erb]: Translation

File: app/views/discussions/show.html.erb [hunk 18]
```diff
@@ -0,0 +1,82 @@
+<% content_for :page_title do %>
+    "Discussion"
```
Comment (ritualz) [inline app/views/discussions/show.html.erb]: Translation

File: app/views/discussions/show.html.erb [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+    <!-- TITLE BEGIN -->
+    <div class="block-title clearfix">
+      <div class="block-options pull-right">
+        <% if policy(@trip).update? %>
+            <%= link_to raw('<i class="fa fa-pencil"></i>'),
+                        edit_trip_discussion_path(@trip, @discussion),
+                        id: 'edit-discussion-btn',
+                        style: 'overflow: hidden; position: relative',
+                        class: 'btn btn-effect-ripple btn-warning',
+                        data: {toggle: 'tooltip', original_title: 'Edit'}
```
Comment (ritualz) [inline app/views/discussions/show.html.erb]: Translation

File: app/views/discussions/show.html.erb [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+            %>
+            <%= link_to raw('<i class="fa fa-trash"></i>'),
+                        trip_discussion_path(@trip, @discussion),
+                        method: :delete,
+                        data: {
+                                confirm: "Are you sure you want to delete this discussion?",
```
Comment (ritualz) [inline app/views/discussions/show.html.erb]: Translation

File: app/views/discussions/show.html.erb [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+                        style: 'overflow: hidden; position: relative;' %>
+            </div>
+        <% end %>
+        <div class="block-options pull-left">
+          <%= link_to trip_discussions_path,
+                      id: 'discussion-view-back',
+                      class: 'btn btn-effect-ripple btn-default',
+                      style: 'overflow: hidden; position: relative;' do %>
+              <i class="fa fa-chevron-left"></i>
+              Back to Discussions
```
Comment (ritualz) [inline app/views/discussions/show.html.erb]: Translation

File: app/views/discussions/show.html.erb [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+    </ul>
+    <ul class="media-list push">
+      <li class="media">
+        <%= link_to image_tag(current_user.gravatar_url, class: 'img-circle', alt: 'Avatar'),
+                    user_path(current_user),
+                    class: 'pull-left'
+        %>
+        <div class="media-body">
+          <%= form_for [@trip, @discussion, @comment], remote: true, id: 'new_comment' do |f| %>
+              <%= f.text_area :body, placeholder: 'Write your comment..', rows: '2', class: 'form-control' %>
```
Comment (ritualz) [inline app/views/discussions/show.html.erb]: Translation

File: app/views/discussions/show.html.erb [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+              <%= f.submit "Publish", id: 'save-comment-btn', class: 'btn btn-effect-ripple btn-sm btn-primary', style: 'overflow: hidden; position: relative;' %>
```
Comment (ritualz) [inline app/views/discussions/show.html.erb]: Translation

File: app/views/users/show.html.erb [hunk 24]
```diff
@@ -57,9 +57,11 @@
   <div class="col-md-7 col-lg-8">
     <div class="block full">
       <div class="block-title">
+        <% if @current_user == @user %>
         <div class="block-options pull-right">
           <a href="<%= edit_user_path(@user) %>" class="btn btn-effect-ripple btn-default" data-toggle="tooltip" title="Edit user"><i class="fa fa-cog"></i></a>
```
Comment (ritualz) [inline app/views/users/show.html.erb]: Translation

File: spec/features/user_page_spec.rb [hunk 25]
```diff
@@ -14,23 +14,17 @@
 
   before do
     sign_in_as!(user)
-    visit user_path(user)
+    visit "/users/#{user.id}"
```
Comment (fsauter) [inline spec/features/user_page_spec.rb]: Why override the user_path?

File: spec/features/user_page_spec.rb [hunk 26]
```diff
@@ -76,4 +70,11 @@
 
     expect(page).to have_content('Email has already been taken')
   end
+
+  scenario "can't edit other peoples user pages" do
+    other_user = FactoryGirl.create(:user)
+    visit "/users/#{other_user.id}"
```
Comment (fsauter) [inline spec/features/user_page_spec.rb]: use user_path?

File: spec/features/user_page_spec.rb [hunk 27]
```diff
@@ -76,4 +70,11 @@
...
+
+    expect(page).to_not have_selector(:link_or_button, "Edit")
```
Comment (fsauter) [inline spec/features/user_page_spec.rb]: use id or translation in tets

File: spec/mailers/forgot_password_spec.rb [hunk 28]
```diff
@@ -0,0 +1,38 @@
+require 'rails_helper'
+
+RSpec.feature "I forgot the password" do
+  let!(:forgetful_user) { FactoryGirl.create(:user, email: "iforget@forget.com") }
+
+  before do
+    ActionMailer::Base.deliveries.clear
+    visit '/'
```
Comment (fsauter) [inline spec/mailers/forgot_password_spec.rb]: use root_path

File: spec/mailers/forgot_password_spec.rb [hunk 29]
```diff
@@ -0,0 +1,38 @@
...
+
+    click_link 'Sign In'
```
Comment (fsauter) [inline spec/mailers/forgot_password_spec.rb]: Use element ids!

File: spec/mailers/forgot_password_spec.rb [hunk 30]
```diff
@@ -0,0 +1,38 @@
...
+    click_link 'forgot_password'
+    fill_in "user_email", with: 'iforget@forget.com'
+    click_button "Remind Password"
```
Comment (fsauter) [inline spec/mailers/forgot_password_spec.rb]: element id

File: spec/mailers/test_mail_sent_spec.rb [hunk 31]
```diff
@@ -0,0 +1,16 @@
...
+  let!(:user) { FactoryGirl.create(:user)}
+
+  before do
+    visit new_user_password_path
+  end
+
+  it "sends email to ActionMailer::Base.deliveries" do
+    expect {
+      fill_in "user_email", with: user.email
+      click_button "Remind Password"
```
Comment (fsauter) [inline spec/mailers/test_mail_sent_spec.rb]: element id

**Comment by fsauter** [comment]
Nice coding! Btw the tests are failing: Neither Pony nor ActionMailer appear to be loaded so email-spec is requiring ActionMailer. -> Seems to be an Travis related error but we should wait with merging until we can solve this because otherwise all new merge requests based on this will fail as well.

**Comment by runarf** [comment]
Yep I will fix it asap. Thanks for the feedback guys 👍

**Comment by fsauter** [comment]
Cool, looking forward to the discussion forum :hamburger: And I think you have to do another rebase so we can merge automatically :+1:

**Comment by fsauter** [comment]
after that i can update the view, the theme was updated and includes now a cool forum view :smiley:

File: config/environments/development.rb [hunk 32]
```diff
@@ -40,4 +40,30 @@
...
+  #    domain: ENV["GMAIL_DOMAIN"],
+  #    authentication: "plain",
+  #    enable_starttls_auto: true,
+  #    user_name: ENV["GMAIL_USERNAME"],
+  #    password: ENV["GMAIL_PASSWORD"]
+  #}
+
+  # mailgun configuration
+  config.action_mailer.smtp_settings = {
+      address: "smtp.mailgun.org",
```
Comment (fsauter) [inline config/environments/development.rb]: We should remember to set address and port also to ENV variables.

**Pull request merged**